experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,311 @@
+        self as u32
+    }
+}
+
+impl ValueWidth for u16 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits16;
+    const ERROR_VALUE_RUST: u16 = u16::MAX;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u32 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits32;
+    const ERROR_VALUE_RUST: u32 = u32::MAX;
+
+    fn cast_to_widest(self) -> u32 {
+        self
+    }
+}
+
+// TrieType trait
+
+/// A trait representing the "trie type" of a [`CodePointTrie`].
+///
+/// Currently, the options are "fast" and "small", which differ in the "fast max"
+/// limit.
+pub trait TrieType {
+    const FAST_MAX: u32;
+    const ENUM_VALUE: TrieTypeEnum;
+}
+
+/// An empty struct to represent "fast" type code point tries for the
+///  [`TrieType`] trait. The "fast max" limit is set to 0xffff.
+pub struct Fast;
+
+impl TrieType for Fast {
+    // All code points up to the fast max limit are represented
+    /// individually in the `index` array to hold their `data` array position, and
+    /// thus only need 2 lookups for a [`crate::codepointtrie::CodePointTrie::get`].
+    /// Code points above the "fast max" limit require 4 lookups.
+    const FAST_MAX: u32 = FAST_TYPE_FAST_INDEXING_MAX;
+    /// This enum variant represents the specific instance of `TrieType` such
+    /// that the enum discriminant values matches ICU4C's enum integer value.
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+/// An empty struct to represent "small" type code point tries for the
+///  [`TrieType`] trait. The "fast max" limit is set to 0x0fff.
+pub struct Small;
+
+impl TrieType for Small {
+    const FAST_MAX: u32 = SMALL_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+/// This struct represents a de-serialized CodePointTrie that was exported from
+/// ICU binary data.
+///
+/// For more information:
+/// - [ICU Site design doc](http://site.icu-project.org/design/struct/utrie)
+/// - [ICU User Guide section on Properties lookup](https://unicode-org.github.io/icu/userguide/strings/properties.html#lookup)
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+/// This struct contains the fixed-length header fields of a [`CodePointTrie`].
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    /// Returns a new [`CodePointTrie`] backed by borrowed data for the `index`
+    /// array and `data` array, whose data values have width `W`, for a trie
+    /// type `T`.
+    pub fn try_new(
+        header: CodePointTrieHeader,
+        index: ZeroVec<'trie, u16>,
```

Comment:
Question: should the type be `T` instead of `u16`?

Reply:
No. The index array always contains uint16_t entries.